File: frontend/src/services/auth-service.ts
import { apiClient } from "@/lib/api-client";
import { unwrap } from "@/lib/api-helpers";
import type { ApiResponse } from "@/types/api";
import type {
  AuthResponse,
  ForgotPasswordRequest,
  GoogleLoginRequest,
  LoginRequest,
  LoginResponse,
  RegisterCompanyRequest,
  RegisterStudentRequest,
  ResendVerificationRequest,
  ResetPasswordRequest,
} from "@/types/auth";
import { tokenStorage } from "@/lib/token-storage";

/** Thin, typed wrapper over the /api/v1/auth endpoints. Unwraps the ApiResponse envelope. */
export const authService = {
  async login(payload: LoginRequest): Promise<LoginResponse> {
    const { data } = await apiClient.post<ApiResponse<LoginResponse>>("/auth/login", payload);
    return unwrap(data);
  },

  /** Exchanges a Google ID token for our own access/refresh token pair. */
  async googleLogin(payload: GoogleLoginRequest): Promise<LoginResponse> {
    const { data } = await apiClient.post<ApiResponse<LoginResponse>>("/auth/google", payload);
    return unwrap(data);
  },

  async registerStudent(payload: RegisterStudentRequest): Promise<AuthResponse> {
    const { data } = await apiClient.post<ApiResponse<AuthResponse>>("/auth/register/student", payload);
    return unwrap(data);
  },

  async registerCompany(payload: RegisterCompanyRequest): Promise<AuthResponse> {
    const { data } = await apiClient.post<ApiResponse<AuthResponse>>("/auth/register/company", payload);
    return unwrap(data);
  },

  async forgotPassword(payload: ForgotPasswordRequest): Promise<string> {
    const { data } = await apiClient.post<ApiResponse<void>>("/auth/forgot-password", payload);
    return data.message;
  },

  async resetPassword(payload: ResetPasswordRequest): Promise<string> {
    const { data } = await apiClient.post<ApiResponse<void>>("/auth/reset-password", payload);
    return data.message;
  },

  async resendVerification(payload: ResendVerificationRequest): Promise<string> {
    const { data } = await apiClient.post<ApiResponse<void>>("/auth/resend-verification", payload);
    return data.message;
  },

  async verifyEmail(token: string): Promise<string> {
    const { data } = await apiClient.get<ApiResponse<{ message: string }>>("/auth/verify-email", {
      params: { token },
    });
    return data.data?.message ?? data.message;
  },

  /** Logs out server-side (revokes the refresh token) then clears local tokens. Best-effort. */
  async logout(): Promise<void> {
    const refreshToken = tokenStorage.getRefreshToken();
    try {
      if (refreshToken) {
        await apiClient.post<ApiResponse<void>>("/auth/logout", { refreshToken });
      }
    } finally {
      tokenStorage.clear();
    }
  },
};

File: src/main/java/com/campusconnect/portal/service/auth/AuthService.java
package com.campusconnect.portal.service.auth;

import com.campusconnect.portal.dto.auth.AuthResponse;
import com.campusconnect.portal.dto.auth.GoogleLoginRequest;
import com.campusconnect.portal.dto.auth.LoginChallengeResponse;
import com.campusconnect.portal.dto.auth.LoginRequest;
import com.campusconnect.portal.dto.auth.LoginResponse;
import com.campusconnect.portal.dto.auth.RefreshTokenResponse;
import com.campusconnect.portal.dto.auth.RegisterCompanyRequest;
import com.campusconnect.portal.dto.auth.RegisterStudentRequest;

/**
 * Authentication and account-lifecycle operations: self-registration (student, company),
 * login with JWT issuance, refresh-token rotation, logout, email verification, and the
 * password-reset flow.
 *
 * <p>Registration does not log the user in — accounts stay disabled until the emailed
 * verification link is followed.
 */
public interface AuthService {

    /**
     * Registers a new student account and its profile, then emails a verification link.
     * The account is created disabled until verification.
     */
    AuthResponse registerStudent(RegisterStudentRequest request);

    /**
     * Registers a new company account and its profile (status {@code PENDING} admin approval),
     * then emails a verification link.
     */
    AuthResponse registerCompany(RegisterCompanyRequest request);

    /**
     * First step of two-factor login: verifies credentials (and that the account is verified,
     * enabled, and unlocked), then emails a one-time code. No tokens are issued here — the client
     * completes the login via {@link #verifyLoginOtp}.
     *
     * @return a challenge describing that an OTP was emailed
     */
    LoginChallengeResponse login(LoginRequest request, String userAgent, String ipAddress);

    /**
     * Second step of two-factor login: consumes the emailed OTP for the given email and issues an
     * access/refresh token pair. Re-checks account state in case it changed since step one.
     *
     * @param userAgent originating User-Agent header (for token audit); may be {@code null}
     * @param ipAddress originating client IP (for token audit); may be {@code null}
     */
    LoginResponse verifyLoginOtp(String email, String code, String userAgent, String ipAddress);

    /**
     * Re-issues a login OTP for an in-progress two-step login. Silent (no error) if the email is
     * unknown or ineligible, to avoid disclosing account state.
     */
    void resendLoginOtp(String email);

    /**
     * Authenticates via a verified Google ID token and issues an access/refresh token pair.
     * A first-time Google email provisions a new enabled STUDENT account (email pre-verified,
     * no profile yet); the user completes their profile afterwards.
     *
     * @param userAgent originating User-Agent header (for token audit); may be {@code null}
     * @param ipAddress originating client IP (for token audit); may be {@code null}
     */
    LoginResponse loginWithGoogle(GoogleLoginRequest request, String userAgent, String ipAddress);

    /**
     * Rotates a refresh token, returning a new access/refresh pair and invalidating the old
     * refresh token.
     */
    RefreshTokenResponse refresh(String refreshToken, String userAgent, String ipAddress);

    /** Logs out by deleting the presented refresh token. Idempotent. */
    void logout(String refreshToken);

    /** Consumes an email-verification token, enabling the account. */
    void verifyEmail(String token);

    /** Re-issues a verification link for an unverified account. Silent if already verified. */
    void resendVerification(String email);

    /**
     * Starts the password-reset flow by emailing a reset link. Always succeeds from the
     * caller's perspective (no account enumeration), regardless of whether the email exists.
     */
    void forgotPassword(String email);

    /** Completes a password reset: consumes the token, sets the new password, revokes sessions. */
    void resetPassword(String token, String newPassword);
}
